Signal/src/util/DisplayableText.swift

```diff
@@ -4,20 +4,330 @@
 
 import Foundation
 
+extension UnicodeScalar {
+    class EmojiRange {
+        // rangeStart and rangeEnd are inclusive.
+        let rangeStart: UInt32
+        let rangeEnd: UInt32
+
+        // MARK: Initializers
+
+        init(rangeStart: UInt32, rangeEnd: UInt32) {
+            self.rangeStart = rangeStart
+            self.rangeEnd = rangeEnd
+        }
+    }
+
+    // From:
+    // https://www.unicode.org/Public/emoji/
+    // Current Version:
+    // https://www.unicode.org/Public/emoji/6.0/emoji-data.txt
+    static let kEmojiRanges = [
```

Comment:
To do binary search, the ranges must be sorted in ascending order.

Reply:
The old emoji detection solution didn't properly count multi-character emoji.  I've decided to be standards-compliant and just do binary search on the ranges defined in the latest emoji standard.  Should be more accurate, precise and performant.

Reply:
I like this new approach. Can you cross reference the script which generates these ranges for the future peeps?

Reply:
Done.